### DISCUSSION TITLE
Feature suggestion: Adding calls to external functions in contract construction

### DISCUSSION THREAD

**Discussion by abadithela** [pacti-org/pacti#20] [Ideas]
This is to start a discussion on whether it would be useful to have the ability to call external functions in contract construction. The current version of gear supports contract assumptions and guarantees written as polyhedral constraints. Often, while the assumptions and guarantees of a system component can be expressed as inequality constraints, it is not necessary that the output variables have a clear linear dependence on the input variables. 

Consider the following example from confmatrix branch example simple_car_ped.json. For this example, the output variable $P$ of contract 2 depends on the input $p$ as follows: $P = p**5$. The assumptions and guarantees are linear constraints: $0leq p \leq 1$ and $0.99 \leq P \leq 1$. However, the relation between $P$ and $p$ is also included as a guarantee in contract 2, by approximating $P = p^5$ as $P = 5p-4$ using series expansion. 

We can argue that the relation $P = p^5$ is internal to the component implemented from contract 2, and need not be described as a fully specified guarantee in the contract definition itself. A couple of discussion points arise from this:

1. Can we have a guarantee call an external function? In this example, can we have one of the guarantees as $P = f(p)$, where $f$ is external to the contract definition, and is a function that describes the dependence between $P$ and $p$. This dependence need not be linear as long as Gear is able to repeatedly call this function to understand how $P$ depends on $p$, and use this to synthesize the quotient. This can potentially describe a much richer class of systems but still have the assumptions and guarantees defined as linear inequalities.

2. The details of function $f$ need not be fully described to Gear to return correct bounds of inputs $p$. It is possible that gear cannot return tight bounds without full knowledge of this function. How do we properly account for abstracting the function $f$ such that Gear can still return useful contracts?

**Comment by iincer** [comment]
Now we have two ways to use the tool: CLI and an API. Regarding the discussion points,
1. If we use the API, we can process nonlinear constraints outside of gear and invoke gear on the processed constraints. The question is whether Gear should provide tools for carrying out the abstraction/refinement.
2. This gets into the how we compute the abstraction/refinement of the nonlinear constraint...

**Comment by NicolasRouquette** [comment]
I have encountered a similar situation in the space mission case study where power-related effects (ie guarantees) may depend on the spacecraft's orientation relative to the sun. In the general case, this dependency is nonlinear due to the trigonometric characteristics of the orbit. However, at a given orbit position, we can certainly make a reasonable linear approximation. The question then would be how to pass sufficient context information to such an $f$ -- e.g., orbit position, so as to generate a suitable linear approximation.
